DylibBinder/XmlGenerator.cs

```diff
@@ -0,0 +1,285 @@
+			                      ("isRequired", func.IsRequired), ("isConvenienceInit", func.IsConvenienceInit),
+			                      ("objcSelector", func.ObjcSelector), ("isPossiblyIncomplete", func.IsPossiblyIncomplete));
+
+			WriteParameterLists (func.ParameterLists);
+			WriteGenericParameters (func.GenericParameters);
+			writer.WriteEndElement ();
+		}
+
+		void WriteParameterLists (DBParameterLists parameterLists)
+		{
+			writer.WriteStartElement ("parameterlists");
+			foreach (var parameterList in parameterLists.ParameterLists) {
+				WriteParameterList (parameterList);
+			}
+			writer.WriteEndElement ();
+		}
+
+		void WriteParameterList (DBParameterList parameterList)
+		{
+			writer.WriteStartElement ("parameterlist");
+			WriteAttributeStrings (("index", parameterList.Index));
+			foreach (var parameter in parameterList.Parameters) {
+				WriteParameter (parameter);
+			}
+			writer.WriteEndElement ();
+		}
+
+		void WriteParameter (DBParameter parameter)
+		{
+			if (parameter.IsEmptyParameter)
+				return;
+
+			writer.WriteStartElement ("parameter");
+			WriteAttributeStrings (("index", parameter.Index), ("type", parameter.Type),
+			                      ("publicName", parameter.PublicName), ("privateName", parameter.PrivateName),
+			                      ("isVariadic", parameter.IsVariadic));
+			writer.WriteEndElement ();
+		}
+
+		void WriteProperties (DBTypeDeclaration typeDeclaration)
+		{
+			foreach (var property in typeDeclaration.Properties.Properties) {
+				WriteProp (property);
+				WriteFunc (property.Getter);
+				if (property.Setter != null)
+					WriteFunc (property.Setter);
+			}
+		}
+
+		void WriteProp (DBProperty property)
+		{
+			writer.WriteStartElement ("property");
+			WriteAttributeStrings (("name", property.Name), ("type", property.Type),
+			                      ("isStatic", property.IsStatic), ("accessibility", property.Accessibility),
+			                      ("isDeprecated", property.IsDeprecated), ("isUnavailable", property.IsUnavailable),
+			                      ("isOptional", property.IsOptional), ("storage", property.Storage),
+			                      ("isPossiblyIncomplete", property.IsPossiblyIncomplete));
+			writer.WriteEndElement ();
+		}
+
+		void CloseWriter () {
+			writer.WriteEndDocument ();
+			writer.Close ();
+		}
+
+		void WriteAttributeStrings (params (string name, object value) [] attributes)
+		{
+			foreach (var attribute in attributes) {
+				if (attribute.value is string s)
+					writer.WriteAttributeString (attribute.name, s);
+				else if (attribute.value is TypeKind kind)
+					writer.WriteAttributeString (attribute.name, kind.ToString ().ToLower ());
+				else if (attribute.value is TypeAccessibility access)
+					writer.WriteAttributeString (attribute.name, access.ToString ());
+				else if (attribute.value is OperatorType operatorType)
+					writer.WriteAttributeString (attribute.name, operatorType.ToString ());
+				else {
```

Comment:
You are calling .ToString to all of them, do we really need to do that if? Only diff I see is the ToLower in the first.